### DISCUSSION TITLE
MemberNotNull on parameters and returns

### DISCUSSION THREAD

**Discussion by RikkiGibson** [dotnet/csharplang#4563] [Q&A]
https://github.com/dotnet/csharplang/blob/1b4f6ac09fd0f29593cc8bf96244e2b16e2509cd/meetings/2020/LDM-2020-02-05.md#nullability

Currently, MemberNotNull/MemberNotNullWhen can be applied and checked only on member methods:

```cs
public class C
{
    public string? Prop { get; set; }

    [MemberNotNull(nameof(Prop))]
    public void EnsureProp() { Prop ??= "a"; }
}
```

However there are scenarios, especially when you don't own the type that you want to validate, where it would be useful to have the attribute on parameters or returns:

```cs
public static class Extensions
{
    public static bool HasValue([MemberNotNullWhen(true, nameof(C.Prop))] this C c)
}
```

The LDM notes from February 2020 state:

> We see the usefulness of the original scenario, but the broadening we're not sure on the return on investment. Let's support the original scenario through a new attribute, MemberNotNull, that only has members as valid attribute targets to start.
>
> If we find users still hit significant limitations without the parameter and return type support, we can consider broadening in a future release.

In this discussion I'd like to gather any evidence that users *are* seeing significant limitations from this, and would benefit from the parameter and return type support.

**Comment by nathanpovo** [comment]
In my case, I needed to create an extension method to set a nullable property to something that is definitely not null

```csharp
public static T DoSomething<T>() where T : IMyInterface, new()
{
    T myObject = new T();
    myObject.NullableProperty = "Definitely not null";

    return myObject;
}

...

public interface IMyInterface
{
    string? NullableProperty { get; set; }
}
```

With the current implementation of static analysis I have to either check or forgive the nullability of the return value:

```csharp
var result = DoSomething<MyImplementation>();
var length = result.NullableProperty!.Length;
```

or use an `out` property:

```csharp
public static T DoSomething<T>(out string notNullResult)
    where T : IMyInterface, new()
{
    T myObject = new T();
    myObject.NullableProperty = "Definitely not null";

    notNullResult = myObject.NullableProperty;
    return myObject;
}

...

var result = DoSomething<MyImplementation>(out string notNullResult);
var length = notNullResult.Length;
```

Having an attribute to indicate that the return value's property is definitely not null would make this a bit more elegant. However, I'm not sure how it would work with a generic method; maybe something similar to the [`NotNullIfNotNull`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis#specify-conditional-post-conditions-notnullwhen-maybenullwhen-and-notnullifnotnull) attribute?

```csharp
[return: NotNullIfNotNull(nameof(T.NullableProperty))]
public static T DoSomething<T>() where T : IMyInterface, new()
{
    T myObject = new T();
    myObject.NullableProperty = "Definitely not null";

    return myObject;
}
```

I had asked a question about this on stackoverflow [here](https://stackoverflow.com/questions/67786714/how-to-inform-the-compiler-that-a-generic-method-has-set-the-property-of-its-ret).

**Comment by russpowers** [comment]
This would be very useful to me as well, I have many Entity Framework data loading methods that are guaranteed to fill in nullable properties, but there's no way to express that some properties are non-null. I'm constantly having to use ! on properties that I know are non-null, which seems to defeat a much of the point of nullability analysis.

Example:

```C#
class Person
{
    public int Id { get; private set; }
    public List<Address>? Addresses { get; private set; }
}

public static async Task<Person> LoadPersonFull(int personId)
{
   ....
}
```
I'd like to be able to express that, in the Person returned by this static method, Person.Addresses is non-null.

**Comment by bradacina** [comment]
It would be nice to call a Validate(complexObject) function and after it returns the compiler would know that complexObject.Foo and complexObject.Bar are guaranteed not null (since I already checked in the Validate function)

**Comment by KUTlime** [comment]
Another [question](https://stackoverflow.com/questions/75233192/how-to-indicate-that-a-property-of-a-return-value-from-a-method-is-not-null-beca/75233592#75233592) on SO from me.

This would be a very handy feature and it would make static analysis more robust.

**Comment by ds5678** [comment]
I also would benefit from this. Most of my use cases echo those previously mentioned.

## Creation Methods

```cs
[return: MemberNotNull(nameof(MethodDefinition.CilMethodBody))]
public static MethodDefinition CreateManaged()
{
	MethodDefinition managedMethod = new();
	managedMethod.CilMethodBody = new(managedMethod);
	return managedMethod;
}
```

## Extension Methods for Validation

It is extremely common for me to write extension methods that validate members of the `this` parameter.

```cs
public static bool IsManaged([MemberNotNullWhen(true, nameof(MethodDefinition.CilMethodBody))] this MethodDefinition method)
{
	return method.CilMethodBody is not null;
}
```

However, if extensions come in C#13, that would most likely offer a workaround.

**Comment by IanGraingerGMSL** [comment]
I was looking for this just now - I am returning an object in a test setup method and it will always have a nullable property set - be nice to tell the caller that!

**Comment by KyMback** [comment]
any plans on this?
